include/correction.h

```diff
@@ -153,13 +153,24 @@ class HashPRNG {
 // common internal for Binning and MultiBinning
 enum class _FlowBehavior {value, clamp, error};
 
+using _NonUniformBins = std::vector<double>;
+
+struct _UniformBins {
+   double lower; // lower edge of first bin
+   double upper; // upper edge of last bin
+   std::size_t n; // number of bins
+};
+
 class Binning {
   public:
     Binning(const JSONObject& json, const Correction& context);
     const Content& child(const std::vector<Variable::Type>& values) const;
 
   private:
-    std::vector<std::tuple<double, Content>> bins_;
+    std::variant<_UniformBins, _NonUniformBins> bins_; // bin edges
```

Comment:
performance-wise, I can't say whether the extra pointer indirection would give more of a performance hit than the extra space we take because of the extra ~24(?) bytes.

It might be premature optimization, but if you prefer I can run some performance benchmarks.